### PR TITLE
Add an appropriate fallback to hb_signed_mul_overflows

### LINKED ISSUE #2320: Fix `hb_signed_mul_overflows` fallback
Our `hb_signed_mul_overflows` [fallback](https://github.com/harfbuzz/harfbuzz/blob/96d792ae80c448715ef317c9f69eb30dc3d34a10/src/hb-algs.hh#L656) is incorrect, https://github.com/harfbuzz/harfbuzz/pull/2274#issuecomment-614279121 should be definitely fixed before it makes another user.

### PR DESCRIPTION
**Pull request by ebraminio**
Fixes #2320

using float decreases it conversion accuracy as revealed by `test-var-coords.c` test.

### CODE REVIEW

File: src/hb-algs.hh [hunk 1]
```diff
@@ -645,16 +645,26 @@ hb_unsigned_mul_overflows (unsigned int count, unsigned int size)
   return (size > 0) && (count >= ((unsigned int) -1) / size);
 }
 
+static inline bool
+hb_signed_mul_overflows (int x, int y, int &result)
```
Comment (behdad) [inline src/hb-algs.hh:649]: I would use `signed` instead of `int` here but that's just my new taste.

File: src/hb-algs.hh [hunk 2]
```diff
@@ -645,16 +645,26 @@ hb_unsigned_mul_overflows (unsigned int count, unsigned int size)
...
+{
 /* Right now we only have one use for signed overflow and as it
  * is GCC 5.1 > and clang we don't care about its fallback ATM */
 #ifndef __has_builtin
 # define __has_builtin(x) 0
 #endif
```
Comment (behdad) [inline src/hb-algs.hh:655]: Since this is leaving a general macro definition behind, move it outside this function, into hb.hh.

File: src/hb-algs.hh [hunk 3]
```diff
@@ -645,16 +645,26 @@ hb_unsigned_mul_overflows (unsigned int count, unsigned int size)
...
 #if __has_builtin(__builtin_mul_overflow)
-# define hb_signed_mul_overflows(x, y, result) __builtin_mul_overflow(x, y, &result)
+  return __builtin_mul_overflow(x, y, &result);
 #else
-# define hb_signed_mul_overflows(x, y, result) (result = (x) * (y), false)
+  long long sink = (long long) x * y;
```
Comment (behdad) [inline src/hb-algs.hh:659]: I'd just use `int64_t`.

File: src/hb-algs.hh [hunk 4]
```diff
@@ -645,16 +645,26 @@ hb_unsigned_mul_overflows (unsigned int count, unsigned int size)
...
+  if (INT_MIN < sink && sink < INT_MAX)
```
Comment (behdad) [inline src/hb-algs.hh:660]: `if (sink == (int) sink)` has only one comparison.

File: src/hb-algs.hh (hunk 1 above)
Comment (ebraminio) [reply src/hb-algs.hh:649]: It is understandable, done.

File: src/hb-algs.hh (hunk 2 above)
Comment (ebraminio) [reply src/hb-algs.hh:655]: done

File: src/hb-algs.hh (hunk 3 above)
Comment (ebraminio) [reply src/hb-algs.hh:659]: done

File: src/hb-algs.hh (hunk 4 above)
Comment (ebraminio) [reply src/hb-algs.hh:660]: done

**Comment by behdad** [comment]
> using float decreases it conversion accuracy as revealed by `test-var-coords.c` test.

Can you share your findings?  The int math in avar is really hard to follow right now and to prove non-overflow.

File: src/hb-algs.hh [hunk 5]
```diff
@@ -645,16 +645,21 @@ hb_unsigned_mul_overflows (unsigned int count, unsigned int size)
...
+  return __builtin_mul_overflow(x, y, &result);
 #else
-# define hb_signed_mul_overflows(x, y, result) (result = (x) * (y), false)
+  int64_t sink = (int64_t) x * y;
+  if (sink == (int) sink)
+  {
+    result = sink;
+    return false;
+  }
+  return true;
```
Comment (behdad) [inline src/hb-algs.hh:660]: This looks good.  You can add `likely` as well.  Or even simplify to:
```
result = sink;
return result == sink;
```

Comment (ebraminio) [reply src/hb-algs.hh]: makes sense, done

**Comment by ebraminio** [comment]
> Can you share your findings? The int math in avar is really hard to follow right now and to prove non-overflow.

Sure, this is a diff applied over current state of PR,

```diff
diff --git a/src/hb-ot-var-avar-table.hh b/src/hb-ot-var-avar-table.hh
index 4e61a3bc..66407abd 100644
--- a/src/hb-ot-var-avar-table.hh
+++ b/src/hb-ot-var-avar-table.hh
@@ -89,14 +89,8 @@ struct SegmentMaps : ArrayOf<AxisValueMap>
     if (unlikely (arrayZ[i-1].fromCoord == arrayZ[i].fromCoord))
       return arrayZ[i-1].toCoord;
 
-    int factor;
-    if (hb_int_mul_overflows (arrayZ[i].toCoord - arrayZ[i-1].toCoord,
-                             value - arrayZ[i-1].fromCoord,
-                             factor))
-      return arrayZ[i-1].toCoord;
-
     int denom = arrayZ[i].fromCoord - arrayZ[i-1].fromCoord;
-    return arrayZ[i-1].toCoord + (factor + denom/2) / denom;
+    return arrayZ[i-1].toCoord + ((float) (arrayZ[i].toCoord - arrayZ[i-1].toCoord) * (value - arrayZ[i-1].fromCoord) + denom/2) / denom;
 #undef toCoord
 #undef fromCoord
   }
diff --git a/test/api/test-var-coords.c b/test/api/test-var-coords.c
index b3aef09e..a5b7df1e 100644
--- a/test/api/test-var-coords.c
+++ b/test/api/test-var-coords.c
@@ -50,7 +50,7 @@ test_get_var_coords (void)
   /* Design coords as input */
   float design_coords[] = {206.f, 0};
   hb_font_set_var_coords_design (font, design_coords, 2);
-  g_assert_cmpint ((int) hb_font_get_var_coords_normalized (font, NULL)[0], ==, -16117);
+  g_assert_cmpint ((int) hb_font_get_var_coords_normalized (font, NULL)[0], ==, -16116);
   g_assert_cmpint ((int) hb_font_get_var_coords_design (font, NULL)[0], ==, 206);
 
   for (float weight = 200; weight < 901; ++weight)
@@ -59,7 +59,7 @@ test_get_var_coords (void)
     hb_ot_var_normalize_coords (face, 1, &weight, &normalized);
     hb_font_set_var_coords_normalized (font, &normalized, 1);
     float converted_back = hb_font_get_var_coords_design (font, NULL)[0];
-    // fprintf (stderr, "%f: %d => %f\n", weight, normalized, converted_back);
+    fprintf (stderr, "%f: %d => %f, error: %f\n", weight, normalized, converted_back, converted_back - weight);
     g_assert_true (G_APPROX_VALUE (converted_back, weight, EPSILON));
   }
```

which makes

```
200.000000: -16384 => 200.000000, error: 0.000000
201.000000: -16338 => 201.049805, error: 0.049805
202.000000: -16294 => 202.038574, error: 0.038574
203.000000: -16249 => 203.039551, error: 0.039551
204.000000: -16205 => 204.028320, error: 0.028320
205.000000: -16160 => 205.041504, error: 0.041504
206.000000: -16116 => 206.030273, error: 0.030273
207.000000: -16072 => 207.019043, error: 0.019043
208.000000: -16027 => 208.032227, error: 0.032227
209.000000: -15982 => 209.045410, error: 0.045410
210.000000: -15938 => 210.034180, error: 0.034180
211.000000: -15893 => 211.035156, error: 0.035156
212.000000: -15849 => 212.023926, error: 0.023926
213.000000: -15804 => 213.037109, error: 0.037109
214.000000: -15760 => 214.025879, error: 0.025879
215.000000: -15715 => 215.039062, error: 0.039062
216.000000: -15671 => 216.027832, error: 0.027832
217.000000: -15626 => 217.041016, error: 0.041016
218.000000: -15581 => 218.054199, error: 0.054199
```

our current EPSILON there is 0.05 that's why it fails now, this is errors in current state.

```
200.000000: -16384 => 200.000000, error: 0.000000
201.000000: -16339 => 201.013184, error: 0.013184
202.000000: -16295 => 202.001953, error: 0.001953
203.000000: -16250 => 203.015137, error: 0.015137
204.000000: -16206 => 204.003906, error: 0.003906
205.000000: -16161 => 205.004883, error: 0.004883
206.000000: -16117 => 205.993652, error: -0.006348
207.000000: -16073 => 206.982422, error: -0.017578
208.000000: -16028 => 207.995605, error: -0.004395
209.000000: -15983 => 209.008789, error: 0.008789
210.000000: -15939 => 209.997559, error: -0.002441
211.000000: -15894 => 211.010742, error: 0.010742
212.000000: -15850 => 211.999512, error: -0.000488
213.000000: -15805 => 213.000488, error: 0.000488
214.000000: -15761 => 213.989258, error: -0.010742
215.000000: -15716 => 215.002441, error: 0.002441
216.000000: -15672 => 215.991211, error: -0.008789
217.000000: -15627 => 217.004395, error: 0.004395
218.000000: -15582 => 218.017578, error: 0.017578
```

**Comment by ebraminio** [comment]
Actually, lets upload float patch separately so we can compare easily.

File: src/hb-algs.hh [hunk 6]
```diff
@@ -645,16 +645,17 @@ hb_unsigned_mul_overflows (unsigned int count, unsigned int size)
...
+hb_int_mul_overflows (int x, int y, int &result)
+{
 #if __has_builtin(__builtin_mul_overflow)
-# define hb_signed_mul_overflows(x, y, result) __builtin_mul_overflow(x, y, &result)
+  return __builtin_mul_overflow (x, y, &result);
 #else
-# define hb_signed_mul_overflows(x, y, result) (result = (x) * (y), false)
+  int64_t sink = (int64_t) x * y;
+  result = sink;
+  return result != (int) sink;
```
Comment (behdad) [inline src/hb-algs.hh:656]: No. You broke it by adding a `(int)` cast to sink.  You want result to be up-converted to sink's type, not other way around.

Comment (behdad) [reply src/hb-algs.hh:656]: Just let compiler do things for you................

File: src/hb-ot-var-avar-table.hh [hunk 7]
```diff
@@ -90,9 +90,9 @@ struct SegmentMaps : ArrayOf<AxisValueMap>
       return arrayZ[i-1].toCoord;
 
     int factor;
-    if (hb_signed_mul_overflows (arrayZ[i].toCoord - arrayZ[i-1].toCoord,
-				 value - arrayZ[i-1].fromCoord,
-				 factor))
+    if (hb_int_mul_overflows (arrayZ[i].toCoord - arrayZ[i-1].toCoord,
```
Comment (behdad) [inline src/hb-ot-var-avar-table.hh:93]: Add `unlikely`.

Comment (ebraminio) [reply src/hb-ot-var-avar-table.hh]: Done

**Comment by behdad** [comment]
Thanks.  lgtm.

**Comment by ebraminio** [comment]
Merging this and decide about #2323 also separately.
